### ISSUE TITLE
LinkedIn OAuth fails after upgrade to v3.3.1": 'A required parameter "client_id" is missing'

### ISSUE THREAD

**Issue by gjovanov** [fastify/fastify-oauth2#58]
## 🐛 Bug Report

Yesterday we have upgraded to v3.3.1 and the LinkedIn OAuth stopped working on https://roomler.live. All other OAuths (FB, Gmail, Github) still work.

LinkedIn fails with the following error:

'A required parameter "client_id" is missing'

I tracked down to simple-oauth2's `client.js`:

```javascript
async request(url, params, opts) {
    const requestOptions = new RequestOptions(this.config, params);
    const options = requestOptions.toObject(opts);

    debug('Creating request to: (POST) %s', url);
    debug('Using request options: %j', options);

    const response = await this.client.post(url, options); // FAILS HERE

    return response.payload;
  }
```

during registration of the fastify-oauth2 plugin for LinkedIn OAuth, I debugged and we are passing the the proper client object with id and secret in it.

So not sure if there was some structural change in `fastify-oauth2` or you think it's `simple-oauth2` issue.

## To Reproduce

linkedin-options.js
```js
const oauthPlugin = require('fastify-oauth2')
const config = require('../../../config')
const defaultState = require('./default-state')
module.exports = {
  name: 'linkedin',
  credentials: {
    client: {
      id: process.env.LINKEDIN_ID,
      secret: process.env.LINKEDIN_SECRET
    },
    auth: oauthPlugin.LINKEDIN_CONFIGURATION
  },
  startRedirectPath: '/oauth/login/linkedin',
  callbackUri: `${config.appSettings.env.URL}/@/oauth/callback/linkedin`,
  scope: ['r_emailaddress', 'r_liteprofile'],
  generateStateFunction: () => {
    return defaultState
  },
  checkStateFunction: (state, callback) => {
    require('./default-check-state')(defaultState, state, callback)
  }
}
```

oauth-controller.js
```js

class OAuthController {
  async getOrCreate (request, reply) {
    const type = request.query.type
    const oauthConfig = this[type] // type="linkedin"
    if (!oauthConfig) {
      throw new TypeError(`Unsupported OAuth type: ${type}`)
    }
    const access = await oauthConfig.getAccessTokenFromAuthorizationCodeFlow(request) // FAILS HERE
    ...
  }

```

api.js
```js
await fastify.register(require('fastify-oauth2'), linkedinOptions)
```

## Expected behavior

A proper LinkedIn token should have been returned.

## Your Environment

- *node version*: v14.4.0
- *fastify version*: 3.0.0
- *os*: Win10

**Comment by mcollina** [comment]
Would you like to send a Pull Request to address this issue? Remember to add unit tests.

**Comment by mcollina** [comment]
I don't have a clue unfortunately :(

**Comment by gjovanov** [comment]
OK I was curious to know if there was some structural change in `fastify-oauth2` specific to LinkedIn, that I'm not aware of.

Anyways, I will debug the issue and see if caused by `fastify-oauth2` or `simple-oauth2` and let you know the outcome and potential fix.

**Comment by gjovanov** [comment]
@mcollina based on:

https://stackoverflow.com/questions/53328295/linkedin-oauth-error-a-required-parameter-client-id-is-missing

it seems that linkedin expects client_id being passed as `application/x-www-form-urlencoded` body.

based othe options from:

https://github.com/lelylan/simple-oauth2/blob/master/API.md#options

after I've extended the `credentials` option like this:

```js
credentials: {
    client: {
      id: process.env.LINKEDIN_ID,
      secret: process.env.LINKEDIN_SECRET
    },
    auth: oauthPlugin.LINKEDIN_CONFIGURATION,
    
    // NEWLY ADDED
    options: {
      bodyFormat: 'form',
      authorizationMethod: 'body'
    }
  },
```

linkedin OAuth started working again.

So it's not an issue, it's just a tweaking in the way we are sending the params.

Closing...
